src/Maestro/maestro-angular/src/app/page/channel-graph/channel-graph.component.ts

```diff
@@ -0,0 +1,136 @@
+import { Component, Input, OnChanges, SimpleChanges, AfterContentInit } from '@angular/core';
+import { graphlib, render } from 'dagre-d3';
+import { select } from 'd3';
+
+import { FlowGraph, FlowRef, FlowEdge } from 'src/maestro-client/models';
+import { RepoNamePipe } from 'src/app/pipes/repo-name.pipe';
+import { analyzeAndValidateNgModules } from '@angular/compiler';
+
+function getRepositoryShortName(repo:string): string {
+  return `${RepoNamePipe.prototype.transform(repo)}`;
+}
+function getNodeLabel(node:FlowRef): string {
+  let label = `${getRepositoryShortName(node.repository)}\n`;
+  label = `${label}${node.branch}\n`;
+  return label;
+}
+
+function getNodeTitle(node:FlowRef): string {
+  let official = node.officialBuildTime == 0 ? "No successful runs in the last 7 days" : node.officialBuildTime.toFixed(2);
+  let pr = node.prBuildTime == 0 ? "No successful runs in the last 7 days" : node.prBuildTime.toFixed(2);
+
+  let title = `Repository: ${getRepositoryShortName(node.repository)}\n`;
+  title = `${title}Branch: ${node.branch}\n`;
+  title = `${title}Official Build: ${official}\n`;
+  title = `${title}Dep Flow: ${pr}\n`;
+  title = `${title}Best Case Path Time: ${node.bestCaseTime.toFixed(2)}\n`;
+  title = `${title}Worst Case Path Time: ${node.worstCaseTime.toFixed(2)}`;
+  return title;
+}
+
+function getNodeDescription(node:FlowRef): string {
+  let description = "A node that is not on the longest build path";
+  if (node.onLongestBuildPath) {
+    description = "A node that is on the longest build path";
+  }
+  return description;
+}
+
+function getEdgeDescription(edge:FlowEdge, graph:FlowGraph): string {
+  let from = graph.nodes.find(x => x.id == edge.fromId);
+  let fromRepo = "";
+  let fromBranch = "";
+  if (from) {
```

Comment:
If either of these fails, the message is... a bit confusing. `An edge that connections @ to @`.  Should we have some default that expresses the "I don't know what happened" state more?

Reply:
Ok, so maybe I should make the repos/branches string | undefined and if they are undefined, change the description to reflect that we don't know what is being connect (or just use the node labels).